Add set algebraic functions to HashSet

weinig opened the pull request:
#### 5e42510363919dd0f4ed7fec7fb23539ad9fe124
<pre>
Add set algebraic functions to HashSet
<a href="https://bugs.webkit.org/show_bug.cgi?id=248881">https://bugs.webkit.org/show_bug.cgi?id=248881</a>
rdar://103074259

Reviewed by Darin Adler.

Adds union, intersection and symmetric difference operations to
HashSet both as in-place mutating operations (the ones with the
`form` prefix) and new set creating operations (the ones with
the `With` suffix).

* Source/WTF/wtf/HashSet.h:
(WTF::W&gt;::unionWith const):
(WTF::W&gt;::intersectionWith const):
(WTF::W&gt;::symmetricDifferenceWith const):
(WTF::W&gt;::formUnion):
(WTF::W&gt;::formIntersection):
(WTF::W&gt;::formSymmetricDifference):
* Tools/TestWebKitAPI/Tests/WTF/HashSet.cpp:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/257569@main">https://commits.webkit.org/257569@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b977aae98af663a649a0985a9d03d61bfb606312

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99223 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8431 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32352 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108615 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168865 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/103224 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8983 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85751 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91724 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106549 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104982 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/6783 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90344 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33786 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88603 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21691 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/76655 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/89930 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2311 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23207 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/85759 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2204 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/45607 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28864 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5210 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/7261 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42681 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/88625 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3680 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19830 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/HashSet.h:
```diff
@@ -132,6 +132,33 @@ class HashSet final {
 
+    // Returns a new set with the elements of both this and the given set.
+    template<typename OtherValueArg, typename OtherHashArg, typename OtherTraitsArg, typename OtherTableTraitsArg>
+    HashSet unionWith(const HashSet<OtherValueArg, OtherHashArg, OtherTraitsArg, OtherTableTraitsArg>&) const;
```
The argument doesn’t have to be a `HashSet`. It could really be any collection including a `std::span`. We take no advantage of the fact that the second one is a set.

darinadler commented on Source/WTF/wtf/HashSet.h:
```diff
@@ -364,6 +391,60 @@ inline auto HashSet<T, U, V, W>::takeAny() -> TakeType
+template<typename OtherValueArg, typename OtherHashArg, typename OtherTraitsArg, typename OtherTableTraitsArg>
+inline auto HashSet<T, U, V, W>::intersectionWith(const HashSet<OtherValueArg, OtherHashArg, OtherTraitsArg, OtherTableTraitsArg>& other) const -> HashSet<T, U, V, W>
+{
+    HashSet<T, U, V, W> result;
```
Somehow I had the impression that if you just write `HashSet` here, it does the same thing as `HashSet<T, U, V, W>`, but maybe that was my imagination.

darinadler commented on Source/WTF/wtf/HashSet.h:
```diff
@@ -364,6 +391,60 @@ inline auto HashSet<T, U, V, W>::takeAny() -> TakeType
+inline auto HashSet<T, U, V, W>::intersectionWith(const HashSet<OtherValueArg, OtherHashArg, OtherTraitsArg, OtherTableTraitsArg>& other) const -> HashSet<T, U, V, W>
+{
+    HashSet<T, U, V, W> result;
+    for (const auto& value : *this) {
```
I would have just written `auto&`. The const happens automatically since this is a const member function.

darinadler commented on Source/WTF/wtf/HashSet.h:
```diff
@@ -132,6 +132,33 @@ class HashSet final {
+    // Returns a new set with the elements that are common to both this set and
+    // the given set.
+    template<typename OtherValueArg, typename OtherHashArg, typename OtherTraitsArg, typename OtherTableTraitsArg>
+    HashSet intersectionWith(const HashSet<OtherValueArg, OtherHashArg, OtherTraitsArg, OtherTableTraitsArg>&) const;
```
I wish the argument type could be a bit more general. This works with any collection that has a `contains` operation, like a `ListHashSet` would be fine or a view on the keys of a `HashMap`.

weinig replied:
I can make this a <typename Collection> and add a note that the requirement is that the collection support contains(). Whenever we can start using `concepts` I can add a nicer refinement for a better error message.

darinadler commented on Source/WTF/wtf/HashSet.h:
```diff
@@ -132,6 +132,33 @@ class HashSet final {
+
+    // Adds the elements of the given set to the set.
+    template<typename OtherValueArg, typename OtherHashArg, typename OtherTraitsArg, typename OtherTableTraitsArg>
+    void formUnion(const HashSet<OtherValueArg, OtherHashArg, OtherTraitsArg, OtherTableTraitsArg>&);
```
The argument doesn’t have to be a `HashSet`. It could really be any collection including a `std::span`. We take no advantage of the fact that the argument is a set.

smfr commented on Source/WTF/wtf/HashSet.h:
```diff
@@ -132,6 +132,33 @@ class HashSet final {
+    HashSet intersectionWith(const HashSet<OtherValueArg, OtherHashArg, OtherTraitsArg, OtherTableTraitsArg>&) const;
+
+    // Returns a new set with the elements that are either in this set or in the
+    // given set, but not in both
```
Might be help to say "XOR"

smfr commented on Source/WTF/wtf/HashSet.h:
```diff
@@ -132,6 +132,33 @@ class HashSet final {
+
+    // Adds the elements of the given set to the set.
+    template<typename OtherValueArg, typename OtherHashArg, typename OtherTraitsArg, typename OtherTableTraitsArg>
+    void formUnion(const HashSet<OtherValueArg, OtherHashArg, OtherTraitsArg, OtherTableTraitsArg>&);
```
Not sure I like "form". I suggest `unite()`, `intersect()`, and something.

weinig replied:
I can't think of a way to make this scheme work with symmetricDifference.

darinadler commented on Source/WTF/wtf/HashSet.h:
```diff
@@ -132,6 +132,33 @@ class HashSet final {
+    // Returns a new set with the elements that are either in this set or in the
+    // given set, but not in both
+    template<typename OtherValueArg, typename OtherHashArg, typename OtherTraitsArg, typename OtherTableTraitsArg>
+    HashSet symmetricDifferenceWith(const HashSet<OtherValueArg, OtherHashArg, OtherTraitsArg, OtherTableTraitsArg>&) const;
```
The argument doesn’t have to be a `HashSet`. It could really be any collection including a `std::span`. We take no advantage of the fact that the argument is a set.

darinadler commented on Source/WTF/wtf/HashSet.h:
```diff
@@ -132,6 +132,33 @@ class HashSet final {
+    // Removes the elements of the set that are also in the given set and adds
+    // the members of the given set that are not already in the set.
+    template<typename OtherValueArg, typename OtherHashArg, typename OtherTraitsArg, typename OtherTableTraitsArg>
+    void formSymmetricDifference(const HashSet<OtherValueArg, OtherHashArg, OtherTraitsArg, OtherTableTraitsArg>&);
```
The argument doesn’t have to be a `HashSet`. It could really be any collection including a `std::span`. We take no advantage of the fact that the argument is a set.

darinadler commented on Source/WTF/wtf/HashSet.h:
```diff
@@ -132,6 +132,33 @@ class HashSet final {
+
+    // Removes the elements of this set that aren't also in the given set.
+    template<typename OtherValueArg, typename OtherHashArg, typename OtherTraitsArg, typename OtherTableTraitsArg>
+    void formIntersection(const HashSet<OtherValueArg, OtherHashArg, OtherTraitsArg, OtherTableTraitsArg>&);
```
I wish the argument type could be a bit more general. This works with any collection that has a `contains` operation, like a `ListHashSet` would be fine or a view on the keys of a `HashMap`.

Commit:
Add set algebraic functions to HashSet
https://bugs.webkit.org/show_bug.cgi?id=248881
rdar://103074259

Reviewed by Darin Adler.

Adds union, intersection and symmetric difference operations to
HashSet both as in-place mutating operations (the ones with the
`form` prefix) and new set creating operations (the ones with
the `With` suffix).

* Source/WTF/wtf/HashSet.h:
(WTF::W>::unionWith const):
(WTF::W>::intersectionWith const):
(WTF::W>::symmetricDifferenceWith const):
(WTF::W>::formUnion):
(WTF::W>::formIntersection):
(WTF::W>::formSymmetricDifference):
* Tools/TestWebKitAPI/Tests/WTF/HashSet.cpp:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/257569@main
